doc/dev/crimson/poseidonstore.rst

```diff
@@ -0,0 +1,352 @@
+        For example, 512 Kbytes of data can be atomically written at once without fsync()
+      - Small size (object_info_t, snapset, collection, etc.) can be embed
+      - stage 1
+        WAL (written) --> | TxBegin A | Log Entry | TxEnd A | (if small)
+        Data partition (written)--> | Data blocks | (if large)
+    - 2. Then, append the data to WAL (if large)
+      - stage 2 (Updating object meta and allocation bitmap can be skipped via pre-allocation)
+        WAL --> | TxBegin A | Log Entry | TxEnd A | (if large)
+
+- Read
+  - Use the cached object metadata to find out the data location
+  - If not cached, need to search WAL after checkpoint and Object meta partition to find the 
+    latest meta data
+
+- Flush
+  - Flush WAL entries whenever needed
+
+
+Crash consistency
+------------------
+Large case
+- 1. Crash occurs right after writing Data blocks
+  - Data partition --> | Data blocks |
+  - We don't need to care this case. Data is not alloacted yet in reality. The blocks will be reused.
+- 2. Crash occurs right after WAL
+  - WAL --> | TxBegin A | Log Entry| TxEnd A |
+  - Data partition --> | Data blocks |
+  - Write procedure is completed, so there is no data loss or inconsistent state
+- 3. Crash occurs after a next write
+  - WAL --> | TxBegin A | Log Entry | TxEnd A | TxBegin B | LogEntry |
+  - Abandon failed write and recover previous state that contains Tx A
+
+Small case
+- 1. Crash occurs right after writing WAL
+  - WAL --> | TxBegin A | Log Entry| TxEnd A |
+  - All data has been written
+
+
+Comparison
+----------
+- Best case (pre-allocation)
+  - Only need two writes on both WAL and Data partition without updating Object meta and Data bitmap
+- Worst case (flush happens)
+  - Without embedding, at least three writes are required additionally on Object meta, Data bitmap, and Data blocks
+- WAL needs to be flushed if the WAL is close to full and the WAL entry is snapset, object_info_t, 
+  collection and pg_log
+  - The premise behind this design is OSD can manage the lateset metadta as a single copy. So,
+    appended entires are not to be read
+- Either best of worst case does not produce severe I/O amplification (it produce I/Os, but I/O rate is constant) 
+  unlike LSM-tree DB (proposed design is similar to LSM-tree which has only level-0)
+  
+
+With NVM
+--------
+With NVM, we are able to make use of a different scheme to store the data depending replication
+performance.
+All of the data store NVM first with out-of-place update manner, then flush them to NVMe storage device
+using in-place update scheme. This is feasible when replication latency (> 0.5 ms) is slower than
+NVMe storage performance (< 0.5 ms) because flush can store all buffered data in NVM in batch-manner.
+
+
+Detailed Design 
+===============
+
+
+- Onode lookup
+Upon onode lookup requests
+First, we hash CID (collection ID) to get the hash index of the hash table. Note that the entire hash table can be 
```

Comment:
As I mentioned, you don't want to include the collection ID here as you don't want to have to rehash every onode upon collection split.

Reply:
See below